File: src/main/java/io/scriptor/isa/Operand.java
package io.scriptor.isa;

import io.scriptor.util.IntSet;
import org.jetbrains.annotations.NotNull;

public record Operand(
        @NotNull String label,
        @NotNull Segment[] segments,
        @NotNull IntSet exclude
) {

    public boolean excludes(final int value) {
        return exclude.contains(extract(value));
    }

    public int extract(final int instruction) {
        int result = 0;
        for (final var segment : segments) {
            final var width = (segment.hi() - segment.lo()) + 1;
            final var mask  = (1 << width) - 1;
            final var bits  = (instruction >>> segment.lo()) & mask;
            result |= bits << segment.shift();
        }
        return result;
    }

    @Override
    public @NotNull String toString() {
        return "%s%s!%s".formatted(label, segments, exclude);
    }
}

File: src/main/java/io/scriptor/isa/Operand.kt
package io.scriptor.isa

import io.scriptor.util.Log.format

data class Operand(
    val label: String,
    val segments: List<Segment>,
    val exclude: MutableSet<Int>,
) {
    fun excludes(value: Int): Boolean {
        return decode(value) in exclude
    }

    fun decode(instruction: Int): Int {
        var result = 0
        for (segment in segments) {
            val width = (segment.hi - segment.lo) + 1
            val mask = (1 shl width) - 1
            val bits = (instruction shr segment.lo) and mask
            result = result or (bits shl segment.shift)
        }
        return result
    }

    fun encode(instruction: Int, value: Int): Int {
        var result = instruction
        for (segment in segments) {
            val width = (segment.hi - segment.lo) + 1
            val mask = (1 shl width) - 1
            val bits = (value shr segment.shift) and mask
            result = result and (mask shl segment.lo).inv()
            result = result or (bits shl segment.lo)
        }
        return result
    }

    override fun toString(): String {
        return format("%s%s!%s", label, segments, exclude)
    }
}
